internal/overlord/logstate/forwarder_test.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/servicelog"
+	. "gopkg.in/check.v1"
+)
+
+type forwarderSuite struct{}
+
+var _ = Suite(&forwarderSuite{})
+
+func (s *forwarderSuite) TestForwarder(c *C) {
+	serviceName := "foobar"
+	rb := servicelog.NewRingBuffer(1024)
+	it := rb.HeadIterator(0)
+	w := servicelog.NewFormatWriter(rb, serviceName)
+
+	cl := &fakeLogClient{make(chan []string, 1)}
+	f := newLogForwarderForTest(serviceName, it, cl, 10*time.Millisecond)
+
+	forwardStopped := make(chan struct{})
+	go func() {
+		f.forward()
+		close(forwardStopped)
+	}()
+
+	writeLog := func(logLine string) {
+		_, err := fmt.Fprintln(w, logLine)
+		c.Assert(err, IsNil)
+	}
+	expectLogs := func(request []string) {
+		select {
+		case req := <-cl.requests:
+			c.Assert(req, DeepEquals, request)
+		case <-time.After(1 * time.Second):
+			c.Fatalf("timed out waiting for request %q", request)
+		}
+	}
+
+	writeLog("log line #1")
+	writeLog("log line #2")
+	writeLog("log line #3")
+	expectLogs([]string{"log line #1", "log line #2", "log line #3"})
+
+	writeLog("log line #4")
+	expectLogs([]string{"log line #4"})
+
+	writeLog("log line #5")
+	writeLog("log line #6")
+	expectLogs([]string{"log line #5", "log line #6"})
```

Comment:
Are we going to be testing both the timeout and the buffer full case here? It seems to me we'll just be testing the 10ms timeout and not the buffer full case, because `expectLogs` is waiting. If so, would be good to test both cases.

Reply:
I'll write a separate test for the buffer full case.